Implement BIP 345 validation (OP_VAULT)

jamesob opened the pull request:
The usual cross-port process from https://github.com/bitcoin/bitcoin/pull/26857.

BIP: https://github.com/bitcoin/bips/pull/1421

jamesob committed:
test: make ECKey more easily constructable

jamesob committed:
consensus: deferred check framework

Implements a framework for accumulated "deferred checks" during vin
script execution, and then executing the checks after all inputs have
been validated individually.

This facilitates new forms of validation which involve gathering
information during vin script execution and then using that information
to do "aggregate" checks on the transaction as a whole.

For vaults in particular, this allows us to conjoin operations across
vaults with incompatible parameters into the same transaction, as well
as include unrelated inputs and outputs, which facilitates more flexible
fee management.

There are also applications re: batch validation and cross-input
signature aggregation.

jamesob committed:
test: make TaprootInfo hashable

jamesob committed:
test: add pprint_tx

jamesob committed:
consensus: add OP_VAULT deployment parameters

jamesob committed:
consensus: implement OP_VAULT and OP_VAULT_RECOVER

OP_VAULT adds a native way of doing vaults - that is, encumbering coins
in such a way that they must pass through a delay period before being
successfully spent, aside from a predetermined path (the recovery
path) which they can be swept to at any time during the life of the
vault.

Thanks to feedback from

- John Moffett: explicating recursive eval attack and providing a test
  case.

Co-authored-by: Greg Sanders <gsanders87@gmail.com>
Co-authored-by: Anthony Towns <aj@erisian.com.au>

jamesob committed:
test: add MiniWallet.get_utxo_as_txin

jamesob committed:
test: add OP_VAULT functional tests

jamesob committed:
inquisition: deployment changes for VAULT

ariard commented on src/script/interpreter.cpp:
```diff
@@ -1264,6 +1290,179 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                                recovery_vout_idx, execdata, recovery_params, recovery_spk, flags)) {
+                            return set_error(serror, *err);
+                        }
+                        // If an optional recovery authorization witness program has been given,
```
From my understanding, the optional recovery authorisation witness program is included in the OP_VAULT `recovery_params`. Therefore in case of future script extensions deployed a posteriori of the OP_VAULT commitment, one has to re-spend your old OP_VAULT output to a new one to benefit from script extensions. I think an alternative could be to have a `recovery_pubkey` blessing the new scripts, in the way grafroot envisioned it: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2018-February/015700.html

jamesob replied:
I think graftroot-like key delegation is probably out of scope for this proposal; such a technique would be useful for many other parts of bitcoin, and if it is implemented, it could be used here too since the recovery params are compatible with future witness verisons. In the meantime if you want to change the recovery rules for a given vault, I think withdrawaling it into a new vault is a reasonable solution.

ariard replied:
Sure the recovery params are compatible with future witness versions, however a specific witness version is included as part of the recovery params hash (e.g version 2 or version 3 witness program) iiuc. Therefore if you would like to upgrade your recovery authorization witness program towards a new witness program you have to unvault the funds towards a new vault instance.

Instead, if you have a grafroot-like key delegation you just have to re-sign the surrogate scripts, I think. Of course, it doesn't invalidate the old surrogates that might present now a safety downgrade from your vault policy. However, this solution could be batched to all your vaults under custody, as they can all share the grafroot-like key ? So sounds a trade-off between operational complexity and introducing a synchronicity point if you have to withdraw all the funds to a new vault if you would like to change the recovery rules.

ajtowns commented on src/script/interpreter.cpp:
```diff
@@ -1946,6 +2066,186 @@ bool GenericTransactionSignatureChecker<T>::CheckDefaultCheckTemplateVerifyHash(
+    const auto& vout = this->txTo->vout;
+
+    assert(recovery_vout_idx >= 0);
+    if (vout.size() < static_cast<size_t>(recovery_vout_idx)) {
```
It's also an error if `recovery_vout_idx == vout.size()`

ajtowns replied:
(Should just pass a `size_t` in in the first place, no?)

jamesob committed:
test: allow try_rpc to take multiple possible errors

michaelfolkson commented:
I'd like to (attempt to) follow the iteration of the vault design (that uses OP_VAULT and other new opcodes) and rationale for changes to that vault design. And this perhaps might not get tracked in the OP_VAULT BIP as that is specifying the new opcode(s) with less of an emphasis on the different vault designs using these opcodes.

@instagibbs stated at the Contracting Primitives meeting that the current vault implementation uses:

> OP_VAULT_RECOVER (sweep to cold like func), OP_VAULT (simple OP_TLUV) and a vault would use OP_CTV for final step

Are we using OP_VAULT and OP_TLUV terminology interchangeably now? And where would be best to track the evolution of the actual vault design? Here? The BIP draft? A Bitcoin StackExchange post?

instagibbs commented:
@michaelfolkson it's evolving based on conversations here https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2023-March/021510.html

jamesob commented:
An updated version of the BIP and explanatory post are coming this week.

pinheadmz commented on src/script/interpreter.cpp:
```diff
@@ -1264,6 +1288,114 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    }
+
+                    // Indicates which (if any) of the vouts carries forward
+                    // the value from the vault into the unvault trigger output.
```
did you mean ...?

```suggestion
                    // the value from the vault into the revault trigger output.
```

jamesob replied:
Fixed, thanks.

jamesob commented:
I've pushed [a few improvements](https://github.com/bitcoin-inquisition/bitcoin/compare/9b7f1511be92867fc7e8c1b909b358ff78a60bcb..904d4dad30618232993fe59b7fc46ca11c7100e6), notably
- assignment of BIP number, and
- removal of parity bit grinding.

This changeset is now fully adherent to [the BIP draft](https://github.com/bitcoin/bips/pull/1421) and, I think, is ready for review. For anyone who missed it, a twitter thread describing the major changes in the OP_VAULT design is here: https://twitter.com/jamesob/status/1639019107432513537

instagibbs commented on src/script/interpreter.cpp:
```diff
@@ -1946,6 +2074,180 @@ bool GenericTransactionSignatureChecker<T>::CheckDefaultCheckTemplateVerifyHash(
+        }
+    }
+    // Append the handling of future witness versions with `else if`s here.
+    else {
```
this doesn't match the spec(and is unsafe, otherwise people can just steal from a vault to a future output version that is miner-stealable)

"If the scriptPubKey of triggerOut is not a witness program of the same version and same tapleaf version as the currently executing script, script execution MUST fail and terminate immediately."

jamesob replied:
I think you missed the above line (https://github.com/bitcoin-inquisition/bitcoin/pull/21/files#diff-a0337ffd7259e8c7c9a7786d6dbd420c80abfa1afdb34ebae3261109d9ae3c19R2191): 
```cpp
    if (!is_unvault_output_wit || trigger_witversion < 1 ||
            trigger_witversion != *execdata.m_curr_witversion) {        // <-- check here
```
Nonetheless, this would be a great test case to add!

instagibbs replied:
ok, in that case I think this is dead code, and something like https://github.com/bitcoin-inquisition/bitcoin/pull/21/files#r1224396136 would help make this clear/easier to test

instagibbs commented on src/script/interpreter.cpp:
```diff
@@ -1264,6 +1288,109 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                case OP_VAULT:
+                {
+                    // Only available in Tapscript
+                    if (sigversion == SigVersion::BASE || sigversion == SigVersion::WITNESS_V0) {
```
This should fail anytime it's not tapscript specifically, which then allows you to clean up `CheckVaultTrigger` significantly, as you know precisely which versions you're dealing with (including tapleaf version). 

A new output version or tapleaf version is likely to have very different semantics anyways (aka script replacements)

jamesob replied:
Good point, fixing this (cc @darosior).

jamesob commented:
I've pushed an update that forces the amount being revaulted to be made explicit. This makes the deferred checks simpler, and avoids spurious invalidation when certain kinds of trigger transactions are constructed. For example, the old code would miscount amounts in the deferred checks if two compatible vault inputs were being triggered for withdrawal and one of them cited a revault output but the other didn't.

@ajtowns has constructed more complicated examples with duplicate trigger and revault outputs which should have been legal but were not. I still have to write testcases for these.

jamesob committed:
fixup! trigger/recovery nAmounts can have extra value

jamesob committed:
fixup! test: trigger/recovery extra value

jamesob committed:
test: vault: add WRITE_TXNS

sanket1729 commented on src/script/interpreter.cpp:
```diff
@@ -1264,6 +1280,116 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    popstack(stack);
+
+                    const CScriptNum n_pushes_data{stacktop(-1), fRequireMinimal};
+                    const size_t n_pushes = static_cast<size_t>(n_pushes_data.GetInt64());
```
Is there an indirect reasoning here to why this cast is safe? I would expect a check of `n n_pushes >= 0` here?

jamesob replied:
Good point about the sign check. We know that with default construction arguments, `CScriptNum` will only permit a 4 byte number (`nMaxNumSize`), so from a capacity standpoint this is okay. But checking explicitly for non-negativeness is a good idea.

jamesob replied:
Fixed.

sanket1729 commented on src/script/interpreter.cpp:
```diff
@@ -1264,6 +1280,116 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    popstack(stack);
+
+                    // Needed on stack: n pushes + the 2 vout indices
+                    if (stack.size() < (n_pushes + 2)) {
```
As commented on BIP, I think this should be `n_pushes + 3`?

jamesob replied:
Fixed.

sanket1729 commented on src/script/interpreter.cpp:
```diff
@@ -1264,6 +1280,116 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+
+                    // Indicates which (if any) of the vouts is a "revault" of some
+                    // portion of the value back into the original vault sPK.
+                    std::optional<size_t> revault_vout_idx = GetVoutIdxFromStack(-2);
```
Feedback for the BIP. I think this is a malleability vector if out < 0. I can see a transaction in the fly and set the value to `int32::MIN` decreasing the fee rate of the transaction. I think we should force that this should be -1 no vout is present.

jamesob replied:
Fixed.

sanket1729 commented on src/validation.cpp:
```diff
@@ -1663,6 +1672,65 @@ bool InitScriptExecutionCache(size_t max_size_bytes)
+                auto trigger_amt{trigger_check->amount - revault_amt};
+                Assume(trigger_amt >= 0);
+
+                expected_trigger_out[trigger_check->vout_idx] += trigger_amt;
```
nit (feel free to ignore): Add a comment why this cannot overflow.

This sum is bounded by total bitcoins. Hence, this sum be atmax 21 * 10**6 * 10**8.

jamesob replied:
Added a comment.

sanket1729 reviewed (commented):
Did a partial review. Only touching the consensus parts.

jamesob committed:
fixup! n_pushes offset and revault idx malleability fixes

Ensure that if <revault-out-idx> is negative, it must be -1 -- otherwise
OP_VAULT scripts are malleable.

Co-authored-by: sanket1729 <sanket1729@gmail.com>

jamesob commented:
This has been rebased on [`25.x`](https://github.com/bitcoin-inquisition/bitcoin/tree/25.x).

ajtowns commented on src/test/fuzz/checkqueue.cpp:
```diff
@@ -14,6 +14,9 @@
     bool result = false;
+    std::vector<DeferredCheck> m_deferred_checks;
+
+    DumbCheck() = default;
```
```
test/fuzz/checkqueue.cpp:19:5: error: unused member function 'DumbCheck' [-Werror,-Wunused-member-function]
    DumbCheck() = default;
    ^
1 error generated.
```

jamesob replied:
Fixed, thanks.

ajtowns reviewed (commented):
unit and functional tests seem to pass for me as at 221ca69a510a1e84eee7194236a0fa8a45e5002c merged onto bd2f51160376229140f9c3c695ebe80e686b98bf with the line below commented out

jamesob reviewed (commented):
All comments addressed, thanks @sanket1729 @ajtowns.

jamesob committed:
bench: add OP_VAULT cost aid

Add a benchmark to help determine how to cost OP_VAULT operations.

jamesob committed:
validation: add OP_VAULT validation weight

ajtowns commented on src/script/interpreter.cpp:
```diff
@@ -1964,7 +2298,9 @@ static bool ExecuteWitnessScript(const Span<const valtype>& stack_span, const CS
-            if (IsOpSuccess(opcode)) {
+            if (is_vault_active && (opcode == OP_VAULT || opcode == OP_VAULT_RECOVER)) {
+                continue;
+            } else if (IsOpSuccess(opcode)) {
```
This violates the "each `SCRIPT_VERIFY_foo` flag is a soft-fork" convention (see #10699 and #26439): if you have a correct `OP_VAULT` invocation, it will be invalid with script-verify flags set to `DISCOURAGE_OP_SUCCESS` but valid with them set to `DISCOURAGE_OP_SUCCESS | VAULT`.

I think this should instead be:

```c++
    if (IsOpSuccess(opcode)) {
        if (opcode == OP_VAULT || opcode == OP_VAULT_RECOVER) {
            if (flags & SCRIPT_VERIFY_DISCOURAGE_VAULT) {
                return set_error();
            } else if ((flags & SCRIPT_VERIFY_VAULT) == 0) {
                // retain OP_SUCCESS behaviour if soft-fork hasn't activated
                return set_success(serror);
            }
        } else {
            if (flags & SCRIPT_VERIFY_DISCOURAGE_OP_SUCCESS) {
                return set_error(serror, SCRIPT_ERR_DISCOURAGE_OP_SUCCESS);
            }
            return set_success(serror);
        }
    }
```

jamesob committed:
Merge branch '25.x' into 2023-02-opvault-inq

michaelfolkson commented:
Why was this closed @jamesob?
